File: backend/app/schemas/meeting.py
from datetime import datetime

from pydantic import BaseModel, Field, field_serializer

from app.schemas.serializers import serialize_utc_datetime


class MeetingCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    scheduled_at: datetime | None = None
    duration: int | None = Field(None, ge=1, le=45)
    user_id: int | None = None


class ParticipantResponse(BaseModel):
    id: int
    session_user_id: str
    display_name: str
    joined_at: datetime
    left_at: datetime | None

    model_config = {"from_attributes": True}

    @field_serializer("joined_at", "left_at")
    def serialize_datetimes(self, value: datetime | None) -> str | None:
        return serialize_utc_datetime(value)


class MeetingResponse(BaseModel):
    id: int
    meeting_id: str
    title: str
    description: str | None
    status: str
    created_at: datetime
    scheduled_at: datetime | None
    duration: int | None
    share_link: str
    user_id: int | None = None
    is_permanent: bool = False
    participants: list[ParticipantResponse] = []

    model_config = {"from_attributes": True}

    @field_serializer("created_at", "scheduled_at")
    def serialize_datetimes(self, value: datetime | None) -> str | None:
        return serialize_utc_datetime(value)


class MeetingListItem(BaseModel):
    id: int
    meeting_id: str
    title: str
    description: str | None
    status: str
    created_at: datetime
    scheduled_at: datetime | None
    duration: int | None
    share_link: str
    participant_count: int
    user_id: int | None = None
    is_permanent: bool = False

    model_config = {"from_attributes": True}

    @field_serializer("created_at", "scheduled_at")
    def serialize_datetimes(self, value: datetime | None) -> str | None:
        return serialize_utc_datetime(value)


class MeetingListResponse(BaseModel):
    public_meetings: list[MeetingListItem]
    my_meetings: list[MeetingListItem]


class JoinMeetingRequest(BaseModel):
    display_name: str = Field(..., min_length=1, max_length=255)


class JoinMeetingResponse(BaseModel):
    meeting: MeetingResponse
    session_user_id: str
    participant_id: int

File: frontend/src/types/meeting.ts
export interface Participant {
  id: number;
  session_user_id: string;
  display_name: string;
  joined_at: string;
  left_at: string | null;
}

export interface Meeting {
  id: number;
  meeting_id: string;
  title: string;
  description: string | null;
  created_at: string;
  scheduled_at: string | null;
  duration: number | null;
  share_link: string;
  participants: Participant[];
}

export interface MeetingListItem {
  id: number;
  meeting_id: string;
  title: string;
  description: string | null;
  created_at: string;
  scheduled_at: string | null;
  duration: number | null;
  share_link: string;
  participant_count: number;
}

export interface MeetingCreatePayload {
  title: string;
  description?: string | null;
  scheduled_at?: string | null;
  duration?: number | null;
}

export interface JoinMeetingResponse {
  meeting: Meeting;
  session_user_id: string;
  participant_id: number;
}

export interface RoomParticipant {
  user_id: string;
  display_name: string;
}

export interface RemotePeer {
  userId: string;
  displayName: string;
  stream: MediaStream | null;
}

export interface JoinPreferences {
  displayName: string;
  micOn: boolean;
  cameraOn: boolean;
}

export type SignalingMessageType =
  | "room-state"
  | "user-joined"
  | "user-left"
  | "offer"
  | "answer"
  | "ice-candidate"
  | "leave"
  | "error"
  | "host-mute"
  | "host-video-off"
  | "host-kick"
  | "kicked"
  | "host-changed";

export interface SignalingMessage {
  type: SignalingMessageType;
  user_id?: string;
  target_user_id?: string;
  payload: Record<string, unknown>;
}
